Switch to clock library

eborden opened the pull request:
The `time` library is not ideal for checking duration of operations. It
is heavy handed from a performance perspective and also includes lower
level systems issues where clock resets can mischaracterize actual
elapsed time. Libraries like `criterion` use their own low level `c`
components to check system time, but the `clock` library abstracts this
behavior nicely. For our purposes we can use a monotonic clock to ensure
we are always accurately depicting wall time of an operation.

This additionally swaps `Duration`'s implementation out for an `Integer`
representing nano seconds. This allows easy implementation of numeric
type classes.

eborden committed:
Switch to clock library

The `time` library is not ideal for checking duration of operations. It
is heavy handed from a performance perspective and also includes lower
level systems issues where clock resets can mischaracterize actual
elapsed time. Libraries like `criterion` use their own low level `c`
components to check system time, but the `clock` library abstracts this
behavior nicely. For our purposes we can use a monotonic clock to ensure
we are always accurately depicting wall time of an operation.

This additionally swaps `Duration`'s implementation out for an `Integer`
representing nano seconds. This allows easy implementation of numeric
type classes.

eborden committed:
Use Duration instead of Integer in tests

eborden committed:
Document Duration

pbrisbin commented on tests/Scientist/Experiment/RunSpec.hs:
```diff
@@ -62,7 +62,7 @@ spec = do
         let control = resultDetailsControl rd
         resultControlValue control `shouldBe` A
-        resultControlDuration control `shouldSatisfy` isDurationNear 0.100
+        resultControlDuration control `shouldSatisfy` isDurationNear 100_000_000
```
It seems like we're saying `result.x.duration` will now be reported as whole nanoseconds instead of decimal seconds.

I'm open to that, if there's a trade-off, but in an ideal world I think this falls under compatibility with the GitHub library we're porting. To keep it consistent, this test should pass as-is with us converting the monotonic nanoseconds to decimal seconds at some point before it gets to the end-user.

That would _probably_ look like leaving the `Duration` type as-is, and converting when it's constructed at `Duration.hs#L25`. Would that undermine the goal of this PR?

eborden replied:
There are a few ways to approach this. Let me play around and see what works.

eborden committed:
Use Fixed data type for Duration representation

The `Nano` data type from `Data.Fixed` gives us all the properties we
want:

- Nanosecond precision
- Display in seconds
- Numeric and fractional instances

eborden committed:
Add fromNanoSecs function

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,54 @@
 {-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+
```
```suggestion
```

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,54 @@
 
+-- | Time elapsed in seconds up to nanosecond precision
+--
+-- >> 1.005
+-- 1.005s
+--
+-- >> toNanoSecs 1.005
+-- 1005000000
```
Might as well make these valid doctest syntax, even if we don't set up a runner yet. Also, we prefer whitespace around documentation bodies when present, so it feels less cramped.

```suggestion
-- >>> 1.005
-- 1.005s
--
-- >>> toNanoSecs 1.005
-- 1005000000
--
```

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,54 @@
+-- | Convert to duration from nanoseconds
+--
+-- >> toNanoSecs 0.000001
+-- >> 1000
```
```suggestion
-- | Convert from duration to nanoseconds
--
-- >>> toNanoSecs 0.000001
-- 1000
--
```

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,54 @@
+
+-- | Convert to duration from nanoseconds
+--
+-- >> fromNanoSecs 1000
+-- >> 0.000001s
```
```suggestion
-- >>> fromNanoSecs 1000
-- 0.000001s
--
```

eborden committed:
Add second conversion functions

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,68 @@
+
+-- | Convert from duration to seconds
+--
+-- >> toSecs 0.000001
+-- >> 0.000001
```
```suggestion
-- >>> toSecs 0.000001
-- 0.000001
--
```

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,68 @@
+
+-- | Convert to duration from seconds
+--
+-- >> toSecs 0.000001
+-- >> 0.000001s
```
```suggestion
-- >>> toSecs 0.000001
-- 0.000001s
--
```

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,68 @@
+-- | Convert from duration to nanoseconds
+--
+-- >> toNanoSecs 0.000001
+-- >> 1000
```
```suggestion
-- >>> toNanoSecs 0.000001
-- 1000
--
```

pbrisbin commented on tests/Scientist/Experiment/RunSpec.hs:
```diff
@@ -55,7 +55,7 @@ spec = do
         $ setExperimentTry (pure A)
         $ setExperimentTry (pure A)
-        $ newExperiment "test" (A <$ threadDelay (100 * 1000))
+        $ newExperiment "test" (A <$ threadDelay 100_000)
```
Neat.

eborden committed:
Fix docs

pbrisbin reviewed (approved):
Thanks!

eborden committed:
Support older GHCs without numeric underscores

pbrisbin commented on library/Scientist/Duration.hs:
```diff
@@ -1,21 +1,68 @@
-  ( Duration(..)
+  ( Duration
   , measureDuration
+  , toSecs
+  , fromSecs
+  , toNanoSecs
+  , fromNanoSecs
```
It's totally fine, but just noting that we only _need_ `toSecs` (or equivalent, say, `unDuration` :wink: ). Everything else is unused, forcing construction to go by `measureDuration`. Similarly with all the numeric instances you gave it. If we were releasing this, I might push on that and suggest we keep the interface as small as necessary to avoid breaks later, but we're a ways away from that.

The numeric instances are interesting in themselves -- we're allowing easy creation of negative durations now, which isn't exactly valid right?

eborden replied:
Personally, when I was first attempting to utilize this type, `unDuration` and the lack of numeric instances was confusing. The API that now exists in this PR feels more complete and understandable to me.

pbrisbin replied:
Calling it `DurationSeconds` would've probably helped.

Giving it [`ToMetricValue`](https://hackage.haskell.org/package/datadog-0.3.0.0/docs/Network-StatsD-Datadog.html#t:ToMetricValue) (and perhaps `ToJSON`) may be the way to have an absolutely minimal but useful/understandable interface.

eborden replied:
Allowing these to be manipulated and constructed like regular numbers is a net positive to me. I have no idea what someone will want to do in their `publish` function, so giving them the primitives affords flexibility.

pbrisbin replied:
That's fair. I'm still not blocking, but how about exposing a `durationToSeconds` and numeric instances only?

I think the other functions are unnecessary. You make a good argument for completeness and treating it like a number, but I don't think it qualifies for the otherwise-unused `from/toNanoSecs` and `fromSecs` functions. And I think the name `durationToSeconds` can help on your understandability point.

WDYT?

eborden committed:
Fix doc

eborden committed:
Remove conversion functions

eborden committed:
Remove toNanoSecs example
